package.json

```diff
@@ -5,11 +5,34 @@
     "email": "kirollos@gmail.com",
     "url": "http://kiro.me"
   },
-  "main": "./dist/fuse.common.js",
-  "module": "./dist/fuse.esm.js",
+  "type": "module",
+  "main": "./dist/fuse.cjs",
+  "module": "./dist/fuse.mjs",
   "unpkg": "./dist/fuse.js",
   "jsdelivr": "./dist/fuse.js",
-  "typings": "./dist/fuse.d.ts",
+  "types": "./dist/fuse.d.ts",
+  "exports": {
+    ".": {
+      "types": "./dist/fuse.d.ts",
+      "import": "./dist/fuse.mjs",
+      "require": "./dist/fuse.cjs"
```

Comment:
For each typed entrypoint, you need a `fuse.d.mts` and `fuse.d.cts` instead of trying to use a single declaration file to represent two different module formats—TypeScript doesn’t support that.

I checked this PR out and ran [@arethetypeswrong/cli](https://www.npmjs.com/package/@arethetypeswrong/cli) on it and every entrypoint was reported as [FalseESM](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseESM.md), meaning any TypeScript users not emitting ES modules will be unable to import this. The linked docs explain in more detail.

Another source from [TypeScript docs](https://www.typescriptlang.org/docs/handbook/esm-node.html#packagejson-exports-imports-and-self-referencing):

> It’s important to note that the CommonJS entrypoint and the ES module entrypoint each needs its own declaration file, even if the contents are the same between them. Every declaration file is interpreted either as a CommonJS module or as an ES module, based on its file extension and the `"type"` field of the package.json, and this detected module kind must match the module kind that Node will detect for the corresponding JavaScript file for type checking to be correct. Attempting to use a single .d.ts file to type both an ES module entrypoint and a CommonJS entrypoint will cause TypeScript to think only one of those entrypoints exists, causing compiler errors for users of the package.

Reply:
Ugh. Well that's doable with some copy actions. I'm on a coding hiatus this week so I'll add that next week.

Reply:
I actually don’t think the two files can be identical with the way the mjs and cjs are written. `Fuse` is the default export of the ESM file, but it’s the `module.exports =` of the CJS file. These two kinds of exports require different declaration syntax.

In ESM:

```ts
export default Fuse;
// or just as well, as you have now
export { Fuse as default }
```

In CJS:

```ts
export = Fuse;
```

(This is an outstanding bug in the declarations that I have a PR fixing at #731.)

When you use `export default Fuse` in a CJS module, it implies this JS:

```ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Fuse;
```

but what you actually have is

```ts
module.exports = Fuse;
```

Reply:
Let me bounce an idea off of you for how to fix this with the least amount of duplication.

1. I create a folder src/typings
2. In here I add `fuse.d.ts.template` which has all the content that is duplicated
3. I also create folder called `outputs`
4. In that folder I create files `fuse.d.cts`, `fuse.d.mts`, and `fuse.d.ts`.
5. Each of these files with have start and end markers through comments (similar to https://github.com/skyra-project/discord-components/blob/933f01a876b488d7477585c5395e0e65727229c9/packages/react/README.md?plain=1#L8) for text injection through a script (similar to https://github.com/skyra-project/discord-components/blob/feat/rewrite-to-lit/scripts/inject-readme-markdown.mjs), as well as the appropriate exports for that type of bundle.

Reply:
attw output after commit d43f45d37ec5e0216c642a2c0f506084f21755a8

![image](https://github.com/krisk/Fuse/assets/4019718/b35d1cdf-fca7-4ce3-abc6-79804cf22096)

Reply:
I have also used my test project to go through the scenarios from my original post again and thanks to these changes the typings now work for _all_ scenarios, including the one that didn't work before. Awesome!